misc/logger/src/circularArray.ts

```diff
@@ -0,0 +1,71 @@
+export default class CircularArray {
+  private array: string[];
+  private head: number;
+  private currentSizeInBytes: number;
+  private capacityInBytes: number;
+
+  constructor(capacityInBytes: number) {
+    this.capacityInBytes = capacityInBytes;
+    this.currentSizeInBytes = 0;
+    this.head = 0;
+    this.array = [];
+  }
+
+  private calculateStringSize(item: string): number {
+    return new TextEncoder().encode(item).length;
+  }
+
+  public enqueue(item: string): void {
+    const itemSize = this.calculateStringSize(item);
+    if (itemSize > this.capacityInBytes) {
+      throw new Error("Item size exceeds total capacity.");
+    }
+
+    let insertIdx = this.array.length;
+    while (this.currentSizeInBytes + itemSize > this.capacityInBytes) {
+      const headItem = this.array[this.head];
+      this.currentSizeInBytes -= this.calculateStringSize(headItem);
+      insertIdx = this.head;
+      this.array[this.head] = "";
+      this.head = (this.head + 1) % Math.max(this.array.length, 1); // Avoid modulo 0, was causing bugs
```

Comment:
This seems wrong. Setting `insertIdx` to `head` and then potentially setting `head` to `0` if you are at the end of the array?

Then when you compare `insertIdx >= this.array.length` below it will always be `false`